File: sdk/dart/lib/src/selectors.dart
/// Artifact selection (SPEC.md section 11).
library;

import 'models.dart';

/// The artifact this client should download, or null if none fits.
///
/// An artifact matches when every selector it declares equals the client's
/// value for that key. Keys the artifact does not mention are ignored, and so
/// are extra keys the client offers, which is what lets a publisher add a new
/// dimension without invalidating older clients.
///
/// The consequence worth knowing: a client that does not declare a key an
/// artifact requires does not match it. That is the intended reading of "every
/// declared pair must hold", and it is why a new required selector needs a
/// client release before it can be used.
Artifact? selectArtifact(
  Manifest manifest,
  Map<String, String> clientSelectors,
) {
  Artifact? best;
  for (final artifact in manifest.artifacts) {
    if (!_matches(artifact, clientSelectors)) continue;
    // Lowest id wins. Fixing the tie-break rather than taking the first match
    // is what makes every implementation agree on malformed input; publishing
    // side already refuses two artifacts with identical selectors.
    if (best == null || artifact.id.compareTo(best.id) < 0) best = artifact;
  }
  return best;
}

/// Every artifact matching this client, ordered the same way [selectArtifact]
/// orders them. Useful for diagnosing "why did it pick that one".
List<Artifact> matchingArtifacts(
  Manifest manifest,
  Map<String, String> clientSelectors,
) {
  final matches = [
    for (final artifact in manifest.artifacts)
      if (_matches(artifact, clientSelectors)) artifact,
  ];
  matches.sort((a, b) => a.id.compareTo(b.id));
  return matches;
}

bool _matches(Artifact artifact, Map<String, String> clientSelectors) {
  for (final entry in selectorsToMap(artifact.selectors).entries) {
    if (clientSelectors[entry.key] != entry.value) return false;
  }
  return true;
}

File: sdk/node/src/selectors.ts
/** Artifact selection (SPEC.md section 11). */

import { selectorsToMap, type Artifact, type Manifest } from "./models.js";

/**
 * Whether every key in `required` equals the client's value for that key. An
 * empty `required` map matches every client.
 */
export function matchesSelectors(
  required: Record<string, string>,
  clientSelectors: Record<string, string>,
): boolean {
  for (const [key, value] of Object.entries(required)) {
    if (clientSelectors[key] !== value) return false;
  }
  return true;
}

function matches(
  artifact: Artifact,
  clientSelectors: Record<string, string>,
): boolean {
  return matchesSelectors(selectorsToMap(artifact.selectors), clientSelectors);
}

/**
 * The artifact this client should download, or null if none fits.
 *
 * An artifact matches when every selector it declares equals the client's value
 * for that key. Keys the artifact does not mention are ignored, and so are
 * extra keys the client offers, which is what lets a publisher add a new
 * dimension without invalidating older clients.
 *
 * The consequence worth knowing: a client that does not declare a key an
 * artifact requires does not match it. That is the intended reading of "every
 * declared pair must hold", and it is why a new required selector needs a
 * client release before it can be used.
 */
export function selectArtifact(
  manifest: Manifest,
  clientSelectors: Record<string, string>,
): Artifact | null {
  let best: Artifact | null = null;
  for (const artifact of manifest.artifacts) {
    if (!matches(artifact, clientSelectors)) continue;
    // Lowest id wins. Fixing the tie-break rather than taking the first match
    // is what makes every implementation agree on malformed input; publishing
    // side already refuses two artifacts with identical selectors.
    if (best === null || artifact.id < best.id) best = artifact;
  }
  return best;
}

/**
 * Every artifact matching this client, ordered the same way `selectArtifact`
 * orders them. Useful for diagnosing "why did it pick that one".
 */
export function matchingArtifacts(
  manifest: Manifest,
  clientSelectors: Record<string, string>,
): Artifact[] {
  return manifest.artifacts
    .filter((artifact) => matches(artifact, clientSelectors))
    .sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));
}
